docs/guides/src/main/server/fips.adoc

```diff
@@ -0,0 +1,229 @@
+make sure that Keycloak container also uses BCFIPS jars (instead of BC jars) and proper options when started.
+
+Regarding this, it is best to build your own container image as described in the <@links.server id="containers"/> and tweak it to use BCFIPS etc.
+
+For example in the current directory, you can create sub-directory `files` and add:
+
+* BC FIPS jar files as described above
+* Custom keystore file - named for example `keycloak-fips.keystore.bcfks`
+* Security file `kc.java.security` with added provider for SAML
+
+Then create `Dockerfile` in the current directory similar to this:
+
+.Dockerfile:
+[source, dockerfile]
+----
+FROM quay.io/keycloak/keycloak:latest as builder
+
+ADD files /tmp/files/
+
+WORKDIR /opt/keycloak
+RUN cp /tmp/files/*.jar /opt/keycloak/providers/
+RUN cp /tmp/files/keycloak-fips.keystore.* /opt/keycloak/conf/server.keystore
+RUN cp /tmp/files/kc.java.security /opt/keycloak/conf/
+
+RUN /opt/keycloak/bin/kc.sh build --fips-mode=strict
+
+FROM quay.io/keycloak/keycloak:latest
+COPY --from=builder /opt/keycloak/ /opt/keycloak/
+
+ENTRYPOINT ["/opt/keycloak/bin/kc.sh"]
+----
+
+Then build FIPS optimized docker image and start it as described in the <@links.server id="containers"/>. Assumption is to use arguments as described above when starting the image.
+
+== Migration from non-fips environment
+
+If you previously used Keycloak in the non-fips environment, it is possible to migrate it to FIPS environment including it's data. However there are restrictions and considerations as
+mentioned above. To highlight some of them:
+
+* Make sure all the Keycloak functionality relying on keystores uses only supported keystore types. This differs based on whether strict or non-strict mode is used.
+* Kerberos authentication may not work. If your authentication flow uses `Kerberos` authenticator, this authenticator will be automatically switched to `DISABLED` when migrated to FIPS
+environment. It is recommended to remove any `Kerberos` user storage providers from your realm and disable `Kerberos` related functionality in LDAP providers before switching to FIPS environment.
+
+In addition to the above, it is recommended to doublecheck this before switching to FIPS strict mode:
+
+* Make sure that all the Keycloak functionality relying on keys (EG. realm or client keys) use RSA keys of at least 2048 bits
+
+* Make sure that clients relying on `Signed JWT with Client Secret` use at least 14 characters long secrets (ideally generated secrets)
+
+* Password length restriction as described above. In case your users have shorter passwords, it is needed to start the server with the max padding length set to 14 of PBKDF2 provider as mentioned
+above. If you prefer to avoid this option, you can for instance ask all your users to reset their password (for example by `Forget password` link) during first authentication in the new environment.
```

Comment:
```suggestion
earlier. If you prefer to avoid this option, you can for instance ask all your users to reset their password (for example by the `Forget password` link) during the first authentication in the new environment.
```